pkg/kubelet/kubelet_stats.go

```diff
@@ -0,0 +1,644 @@
+			// response.
+			glog.Errorf("Partial failure issuing cadvisor.ContainerInfoV2: %v", err)
+		} else {
+			return nil, fmt.Errorf("failed to get root cgroup stats: %v", err)
+		}
+	}
+
+	infos = removeTerminatedContainerInfo(infos)
+
+	// Map each container to a pod and update the PodStats with container data.
+	podToStats := map[statsapi.PodReference]*statsapi.PodStats{}
+	for key, cinfo := range infos {
+		// On systemd using devicemapper each mount into the container has an
+		// associated cgroup. We ignore them to ensure we do not get duplicate
+		// entries in our summary. For details on .mount units:
+		// http://man7.org/linux/man-pages/man5/systemd.mount.5.html
+		if strings.HasSuffix(key, ".mount") {
+			continue
+		}
+		// Build the Pod key if this container is managed by a Pod
+		if !isPodManagedContainer(&cinfo) {
+			continue
+		}
+		ref := buildPodRef(&cinfo)
+
+		// Lookup the PodStats for the pod using the PodRef. If none exists,
+		// initialize a new entry.
+		podStats, found := podToStats[ref]
+		if !found {
+			podStats = &statsapi.PodStats{PodRef: ref}
+			podToStats[ref] = podStats
+		}
+
+		// Update the PodStats entry with the stats from the container by
+		// adding it to podStats.Containers.
+		containerName := kubetypes.GetContainerName(cinfo.Spec.Labels)
+		if containerName == leaky.PodInfraContainerName {
+			// Special case for infrastructure container which is hidden from
+			// the user and has network stats.
+			podStats.Network = cadvisorInfoToNetworkStats("pod:"+ref.Namespace+"_"+ref.Name, &cinfo)
+			podStats.StartTime = metav1.NewTime(cinfo.Spec.CreationTime)
+		} else {
+			podStats.Containers = append(podStats.Containers, *cadvisorInfoToContainerStats(containerName, &cinfo, &rootFsInfo, &imageFsInfo))
+		}
+	}
+
+	// Add each PodStats to the result.
+	result := make([]statsapi.PodStats, 0, len(podToStats))
+	for _, podStats := range podToStats {
+		// Lookup the volume stats for each pod.
+		podUID := types.UID(podStats.PodRef.UID)
+		if vstats, found := self.resourceAnalyzer.GetPodVolumeStats(podUID); found {
+			podStats.VolumeStats = vstats.Volumes
+		}
+		result = append(result, *podStats)
+	}
+
+	return result, nil
+}
+
+// ImageFsStats returns the stats (from cAdvisor) of the filesystem for storing
+// images.
+func (self *cadvisorStatsProvider) ImageFsStats() (*statsapi.FsStats, error) {
+	imageFsInfo, err := self.cadvisor.ImagesFsInfo()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get imageFs info: %v", err)
+	}
+	imageStats, err := self.imageService.ImageStats()
+	if err != nil || imageStats == nil {
+		return nil, fmt.Errorf("failed to get image stats: %v", err)
+	}
+
+	var imageFsInodesUsed *uint64
+	if imageFsInfo.Inodes != nil && imageFsInfo.InodesFree != nil {
+		imageFsIU := *imageFsInfo.Inodes - *imageFsInfo.InodesFree
+		imageFsInodesUsed = &imageFsIU
+	}
+
+	// Get the root container stats's timestamp, which will be used as the
+	// imageFs stats timestamp.
```

Comment:
https://github.com/google/cadvisor/pull/1724 added the timestamp into the FsInfo returned from cadvisor. In this way, we will not need this separate call to get the timestamp anymore. I will update this PR when the dependency is updated.